src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -770,2524 +369,510 @@ MulticopterPositionControl::task_main_trampoline(int argc, char *argv[])
-		const matrix::Vector2f &unit_current_to_next)
-{
-
-	/* minimum cruise speed when passing waypoint */
-	float min_cruise_speed = 1.0f;
-
-	/* make sure that cruise speed is larger than minimum*/
-	if ((get_cruising_speed_xy() - min_cruise_speed) < SIGMA_NORM) {
-		return get_cruising_speed_xy();
+		_states.position(0) = _states.position(1) = NAN;
 	}
 
-	/* middle cruise speed is a number between maximum cruising speed and minimum cruising speed and corresponds to speed at angle of 90degrees
-	 * it needs to be always larger than minimum cruise speed */
-	float middle_cruise_speed = _cruise_speed_90.get();
+	if (PX4_ISFINITE(_local_pos.z) && _local_pos.z_valid) {
+		_states.position(2) = _local_pos.z;
 
-	if ((middle_cruise_speed - min_cruise_speed) < SIGMA_NORM) {
-		middle_cruise_speed = min_cruise_speed + SIGMA_NORM;
+	} else {
+		_states.position(2) = NAN;
 	}
 
-	if ((get_cruising_speed_xy() - middle_cruise_speed) < SIGMA_NORM) {
-		middle_cruise_speed = (get_cruising_speed_xy() + min_cruise_speed) * 0.5f;
-	}
+	if (PX4_ISFINITE(_local_pos.vx) && PX4_ISFINITE(_local_pos.vy) && _local_pos.v_xy_valid) {
+		_states.velocity(0) = _local_pos.vx;
+		_states.velocity(1) = _local_pos.vy;
+		_states.acceleration(0) = _vel_x_deriv.update(-_states.velocity(0));
+		_states.acceleration(1) = _vel_y_deriv.update(-_states.velocity(1));
 
-	/* if middle cruise speed is exactly in the middle, then compute
-	 * vel_close linearly
-	 */
-	bool use_linear_approach = false;
+	} else {
+		_states.velocity(0) = _states.velocity(1) = NAN;
+		_states.acceleration(0) = _states.acceleration(1) = NAN;
 
-	if (((get_cruising_speed_xy() + min_cruise_speed) * 0.5f) - middle_cruise_speed < SIGMA_NORM) {
-		use_linear_approach = true;
+		// since no valid velocity, update derivate with 0
+		_vel_x_deriv.update(0.0f);
+		_vel_y_deriv.update(0.0f);
 	}
 
-	/* angle = cos(x) + 1.0
-	 * angle goes from 0 to 2 with 0 = large angle, 2 = small angle:   0 = PI ; 2 = PI*0 */
-	float angle = 2.0f;
-
-	if (unit_current_to_next.length() > SIGMA_NORM) {
-		angle = unit_current_to_next * (unit_prev_to_current * -1.0f) + 1.0f;
-	}
+	if (MPC_ALT_MODE.get() && _local_pos.dist_bottom_valid && PX4_ISFINITE(_local_pos.dist_bottom_rate)) {
+		// terrain following
+		_states.velocity(2) = -_local_pos.dist_bottom_rate;
+		_states.acceleration(2) = _vel_z_deriv.update(-_states.velocity(2));
 
-	/* compute velocity target close to waypoint */
-	float vel_close;
+	} else if (PX4_ISFINITE(_local_pos.vz)) {
 
-	if (use_linear_approach) {
+		if (PX4_ISFINITE(vel_sp_z) && fabsf(vel_sp_z) > FLT_EPSILON && PX4_ISFINITE(_local_pos.z_deriv)) {
+			// A change in velocity is demanded. Set velocity to the derivative of position
+			// because it has less bias but blend it in across the landing speed range
+			float weighting = fminf(fabsf(vel_sp_z) / _land_speed.get(), 1.0f);
+			_states.velocity(2) = _local_pos.z_deriv * weighting + _local_pos.vz * (1.0f - weighting);
```

Comment:
@Stifael @MaEtUgR Aren't you overwriting this again a few lines below?

Reply:
fixed